crates/lfan/src/cache/ttl_cache.rs

```diff
@@ -0,0 +1,146 @@
+use std::hash::Hash;
+use std::{borrow::Borrow, time::Instant};
+use std::{collections::HashMap, time::Duration};
+
+use super::{
+    cache::Cache,
+    policy::{EvictionPolicy, InsertionPolicy},
+};
+
+struct CacheItem<V> {
+    item: V,
+    last_seen: Instant,
+}
+
+impl<V> CacheItem<V> {
+    pub fn is_expired(&self, ttl: &Duration) -> bool {
+        let now = Instant::now();
+        &now.duration_since(self.last_seen) > ttl
+    }
+}
+
+impl<V> From<V> for CacheItem<V> {
+    fn from(item: V) -> Self {
+        CacheItem {
+            item,
+            last_seen: Instant::now(),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct TTLCache<K, V, IP, EP>
+where
+    K: Hash + Eq,
+    IP: Default + InsertionPolicy<K>,
+    EP: Default + EvictionPolicy<K>,
+{
+    data: HashMap<K, CacheItem<V>>,
+
+    ttl: Duration,
+
+    insertion_policy: IP,
+    eviction_policy: EP,
+
+    maximum_size: usize,
+}
+
+impl<K, V, IP, EP> TTLCache<K, V, IP, EP>
+where
+    K: Hash + Eq + std::fmt::Debug,
+    IP: Default + InsertionPolicy<K>,
+    EP: Default + EvictionPolicy<K>,
+{
+    pub fn new(maximum_size: usize, ttl: Duration) -> Self {
+        Self {
+            data: HashMap::new(),
+            ttl,
+            insertion_policy: IP::default(),
+            eviction_policy: EP::default(),
+            maximum_size,
+        }
+    }
+}
+
+impl<K, V, IP, EP> Cache for TTLCache<K, V, IP, EP>
+where
+    K: Hash + Eq + std::fmt::Debug,
+    IP: Default + InsertionPolicy<K>,
+    EP: Default + EvictionPolicy<K>,
+{
+    type Key = K;
+    type Value = V;
+
+    fn insert(&mut self, key: K, value: V) -> (bool, Option<V>) {
+        let is_in_cache = self.data.contains_key(&key);
+
+        if is_in_cache {
+            // Update.
+            self.eviction_policy.on_update(&key);
+            self.data.insert(key, value.into());
+            return (true, None);
+        } else if self.data.len() < self.maximum_size {
+            if self.insertion_policy.should_add(&key) {
+                // Straight insert.
+                self.eviction_policy.on_insert(&key);
+                self.data.insert(key, value.into());
+                return (true, None);
+            }
+        } else {
+            // Get a cache victim.
+            let victim_key = self.eviction_policy.get_victim().unwrap(); // If this panics, there is most likely a bug in the cache.
```

Comment:
> If this panics, there is most likely a bug in the cache.

I agree, but what are the side effects when the cache panics? Will it kill the node? Return a 500?

Reply:
it's going to return a 500. Note that it won't panic with the current LRU policy, this unwrap is most likely to be tripped when developing new policies and won't happen during normal operations